File: C_TopInterviewQuestions-HardCollection/a_array_and_strings/a_product_of_array_except_self.rs
struct Solution {}

impl Solution {
    pub fn product_except_self(nums: Vec<i32>) -> Vec<i32> {
        let mut prods = nums.clone();
        for i in 1..nums.len() {
            prods[i] *= prods[i - 1]
        }
        let mut results = vec![0; nums.len()];
        let mut rev_prod: i32 = 1;
        for (i, num) in nums[1..].iter().enumerate().rev() {
            results[i + 1] = prods[i] * rev_prod;
            rev_prod *= num
        }
        results[0] = rev_prod;
        return results;
    }
}

pub fn test() {
    println!("{}", [24, 12, 8, 6].to_vec() == Solution::product_except_self([1, 2, 3, 4].to_vec()))
}

File: C_TopInterviewQuestions-HardCollection/a_array_and_strings/a_product_of_array_except_self.ts
import { MyArrays } from "../lib.ts"

function productExceptSelf(nums: number[]): number[] {
    const results = new Array(nums.length).fill(1)
    for (let i = 1; i < results.length; i++) {
        results[i] *= results[i - 1] * nums[i - 1]
    }
    let tmp = 1
    for (let i = results.length - 1; i > -1; i--) {
        results[i] *= tmp
        tmp *= nums[i]
    }
    return results
}

console.log(MyArrays.equals([24, 12, 8, 6], productExceptSelf([1, 2, 3, 4])))

